packages/tbas/src/tbas.ts

```diff
@@ -0,0 +1,249 @@
+    walletSeed,
+  })
+  if ("error" in response) {
+    throw new Error(response.error)
+  }
+  console.log("Minted NFT", response.txId, response.tokenId)
+  return response
+}
+
+// Create token bound account from implementation
+export async function createAccount(tokenContract:string, tokenId:string, chainId:string, waitForReceipt=false){
+  console.log('Creating account...')
+  const privateKey = process.env.XDC_WALLET_SECRET || ''
+  if(!privateKey){
+    return { status: 'error', txid:'', address:'', error:'Private key not found' }
+  }
+  try {
+    const client = newClient()
+    const serverAccount = privateKeyToAccount(privateKey as Address)
+    const contract = getContract({
+      client,
+      abi: registryABI,
+      address: registryAddress
+    })
+
+    // Simulate
+    const { request } = await client.public.simulateContract({
+      account: serverAccount,
+      address: registryAddress,
+      abi: registryABI,
+      functionName: 'createAccount',
+      args: [implementationAddress, baseSalt, BigInt(chainId), tokenContract as Address, BigInt(tokenId)],
+    })
+
+    // Send to chain
+    const txid = await client.wallet.writeContract(request)
+    console.log('TXID', txid)
+
+    // Tx receipt
+    if(waitForReceipt){
+      const receipt = await getReceipt(txid)
+      if(!receipt){
+        return { status: 'notfound', txid, address:'', error:'' }
+      }
+      if(receipt.status==='success'){
+        const address = `0x${receipt?.logs[0].data.substr(26,40)}`
+        console.log('TBA', address, receipt.status)
+        return { status: receipt.status, txid, address, error:'' }
+      }
+      return { status: receipt.status, txid, address:'', error:'' }
+    }
+    return { status: 'pending', txid, address:'', error:'' }
+  // biome-ignore lint/suspicious/noExplicitAny: any error
+  } catch(ex:any) {
+    console.error(ex)
+    return { status: 'error', txid:'', address:'', error:ex.message }
+  }
+}
+
+// Get account address from contract
+export async function getAccount(tokenContract:string, tokenId:string, chainId:string){
+  console.log('Getting account...', tokenContract, tokenId, chainId)
+  const client = newClient()
+  const contract = getContract({
+    client: client.public,
+    abi: registryABI,
+    address: registryAddress
+  })
+  const address = await contract.read.account([
+    implementationAddress,
+    baseSalt,
+    BigInt(chainId),
+    tokenContract as Address,
+    BigInt(tokenId)
+  ])
+  console.log('TBA', address)
+  return address
+}
+
+
+export async function newAccount(entity_type:string, entity_id:string){
+  try {
+    const resMint = await mintAccountNFT(entity_id) // mint nft for tba in main 721 contract
+    console.log('NFT', resMint)
+    const tokenId = resMint.tokenId
+    console.log('TokenID', tokenId)
+    // create token bound account for user in xdc
+    const account_address = await getAccount(tokenContract, tokenId, chainId) // prefetch account address
```

Comment:
why is this getting and then creating an account?

Reply:
We fetch the account without creating it just to know the address, which should match the result of the creation process. Easier than fetching logs from a transaction result